### ISSUE TITLE
[Flaky Test] Controller Common Suite: [It] CheckerWatchdog

### ISSUE THREAD

**Issue by timebertt** [gardener/gardener#5836]
**How to categorize this issue?**

/area testing
/kind flake

**Which test(s)/suite(s) are flaking**:

`Controller Common Suite: [It] CheckerWatchdog`

**CI link**:

https://prow.gardener.cloud/view/gs/gardener-prow/logs/ci-gardener-unit/1511573836791812096
https://prow.gardener.cloud/view/gs/gardener-prow/logs/ci-gardener-unit/1517072393976156160
https://prow.gardener.cloud/view/gs/gardener-prow/logs/ci-gardener-unit/1518522396259127296

**Reason for failure**:

Multiple failures like:
```
Running Suite: Controller Common Suite - /home/prow/go/src/github.com/gardener/gardener/extensions/pkg/controller/common
========================================================================================================================
Random Seed: 1650533752
Will run 18 of 18 specs
••••••••••••
------------------------------
• [FAILED] [0.002 seconds]
CheckerWatchdog [AfterEach]
/home/prow/go/src/github.com/gardener/gardener/extensions/pkg/controller/common/checkerwatchdog_test.go:55
  #Start / #Stop / #AddContext / #RemoveContext / #Result
  /home/prow/go/src/github.com/gardener/gardener/extensions/pkg/controller/common/checkerwatchdog_test.go:59
    should cancel the context returned by AddContext if the checker returns false
    /home/prow/go/src/github.com/gardener/gardener/extensions/pkg/controller/common/checkerwatchdog_test.go:82
  missing call(s) to *mock.MockChecker.Check(is anything) /home/prow/go/src/github.com/gardener/gardener/extensions/pkg/controller/common/checkerwatchdog_test.go:83
  In [AfterEach] at: /home/prow/go/src/github.com/gardener/gardener/extensions/pkg/controller/common/checkerwatchdog_test.go:56
------------------------------
•••••
Ran 18 of 18 Specs in 0.127 seconds
FAIL! -- 17 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestCommon (0.13s)
FAIL
FAIL	github.com/gardener/gardener/extensions/pkg/controller/common	0.453s
```

**Anything else we need to know**:

**Comment by timebertt** [comment]
I've spent quite some time to reproduce, debug and understand this problem. However, so far I haven't been able to.
Hence, https://github.com/gardener/gardener/pull/5873 adds some debug logs to help debug the flake if it occurs again.

**Comment by rfranzke** [comment]
@timebertt It happened again here: https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/5876/pull-gardener-unit/1520017547204759552

**Comment by rfranzke** [comment]
The logs in above build are:

```
• [FAILED] [0.003 seconds]
CheckerWatchdog [AfterEach]
/home/prow/go/src/github.com/gardener/gardener/extensions/pkg/controller/common/checkerwatchdog_test.go:57
  #Start / #Stop / #AddContext / #RemoveContext / #Result
  /home/prow/go/src/github.com/gardener/gardener/extensions/pkg/controller/common/checkerwatchdog_test.go:61
    should cancel the context returned by AddContext if the checker times out
    /home/prow/go/src/github.com/gardener/gardener/extensions/pkg/controller/common/checkerwatchdog_test.go:126
  Begin Captured GinkgoWriter Output >>
    {"level":"debug","ts":1651236036.4442346,"msg":"Starting watchdog"}
    {"level":"debug","ts":1651236036.44455,"msg":"Adding context","key":"foo"}
    {"level":"Level(-2)","ts":1651236036.4446206,"msg":"Stopping timer"}
    {"level":"Level(-2)","ts":1651236036.4447029,"msg":"Resetting timer"}
    {"level":"Level(-2)","ts":1651236036.4447496,"msg":"Result requested"}
    {"level":"info","ts":1651236036.4449112,"msg":"Cancelling all contexts"}
    {"level":"Level(-2)","ts":1651236036.44[495](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/5876/pull-gardener-unit/1520017547204759552#1:build-log.txt%3A495)82,"msg":"Result requested"}
    {"level":"Level(-2)","ts":1651236036.44[499](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/5876/pull-gardener-unit/1520017547204759552#1:build-log.txt%3A499)28,"msg":"Using last result"}
    {"level":"debug","ts":1651236036.44[505](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/5876/pull-gardener-unit/1520017547204759552#1:build-log.txt%3A505)14,"msg":"Removing context","key":"foo"}
    {"level":"debug","ts":16[512](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/5876/pull-gardener-unit/1520017547204759552#1:build-log.txt%3A512)36036.4451535,"msg":"Stopping watchdog"}
  << End Captured GinkgoWriter Output
  missing call(s) to *mock.MockChecker.Check(is anything) /home/prow/go/src/github.com/gardener/gardener/extensions/pkg/controller/common/checkerwatchdog_test.go:127
```

Does this help @timebertt ?

**Comment by rfranzke** [comment]
/assign @timebertt
since he wants to take another look

**Comment by timebertt** [comment]
The debug logs were helpful indeed. I've managed to understand what's going on in these flakes:

In the successful case, the logs show this order of execution:
```
  Begin Captured GinkgoWriter Output >>
    {"level":"debug","ts":1651484912.7265851,"msg":"Starting watchdog"}
    {"level":"debug","ts":1651484912.726625,"msg":"Adding context","key":"foo"}
    {"level":"Level(-2)","ts":1651484912.7266822,"msg":"Stopping timer"}
    {"level":"Level(-2)","ts":1651484912.726752,"msg":"Resetting timer"}
    {"level":"Level(-2)","ts":1651484912.726759,"msg":"Result requested"}
    {"level":"info","ts":1651484912.726826,"msg":"Cancelling all contexts"}
    {"level":"Level(-2)","ts":1651484912.7268338,"msg":"Timer expired"}
    {"level":"info","ts":1651484912.726847,"msg":"Cancelling all contexts"}
    {"level":"Level(-2)","ts":1651484912.7268531,"msg":"Result requested"}
    {"level":"Level(-2)","ts":1651484912.726857,"msg":"Using last result"}
    {"level":"debug","ts":1651484912.726868,"msg":"Removing context","key":"foo"}
    {"level":"Level(-2)","ts":1651484912.726872,"msg":"Stopping timer"}
    {"level":"debug","ts":1651484912.726876,"msg":"Stopping watchdog"}
    {"level":"Level(-2)","ts":1651484912.726882,"msg":"Context done"}
  << End Captured GinkgoWriter Output
```

I.e., `Result requested` -> `Timer expired` -> `Result requested` (`Using last result`)

In the flake however, it looks like this:
```
  Begin Captured GinkgoWriter Output >>
    {"level":"debug","ts":1651236036.4442346,"msg":"Starting watchdog"}
    {"level":"debug","ts":1651236036.44455,"msg":"Adding context","key":"foo"}
    {"level":"Level(-2)","ts":1651236036.4446206,"msg":"Stopping timer"}
    {"level":"Level(-2)","ts":1651236036.4447029,"msg":"Resetting timer"}
    {"level":"Level(-2)","ts":1651236036.4447496,"msg":"Result requested"}
    {"level":"info","ts":1651236036.4449112,"msg":"Cancelling all contexts"}
    {"level":"Level(-2)","ts":1651236036.4449582,"msg":"Result requested"}
    {"level":"Level(-2)","ts":1651236036.4449928,"msg":"Using last result"}
    {"level":"debug","ts":1651236036.4450514,"msg":"Removing context","key":"foo"}
    {"level":"debug","ts":1651236036.4451535,"msg":"Stopping watchdog"}
  << End Captured GinkgoWriter Output
```

I.e., `Result requested` -> `Result requested` (`Using last result`)
Note, that `Timer expired` is not logged here.

This explains the cause of the flake:
Both `case <-w.resultChan:` and `case <-timer.C():` in the [watchdog's `select` statement](https://github.com/gardener/gardener/blob/e0654f04e0c87e87f2f5af7639ebc00e4b24704d/extensions/pkg/controller/common/checkerwatchdog.go#L104-L115) can be unblocked/ready at the same time (same `select` execution) because the test code advances the fake clock and calls `Result` in quick succession (in a different goroutine). If that happens, a random case is chosen, as the `select` statement is not deterministic.

If the "timer expired" case is executed first, then there is a second `checker.Check()` call (expected by the mock) and after that the "result requested case is executed, which at the end makes the `watchdog.Result()` call return as expected by the test code. -> Test succeeds
If the "result requested" case is executed first, the `watchdog.Result()` call returns immediately without a new `Result` call and the test case (`It`) returns. There might not have been a `checker.Check()` call yet, which makes the mock fail the test case. Note however, that the watchdog's goroutine might still be running, so the `Check()` call is actually done, though too late for the mock's expectations.

In short, I would say that the root issue here is that we are testing asynchronous code (watchdog runs in another goroutine) with synchronous test code and mechanisms (mocks don't wait for calls to be done before returning from the test case).
I fixed this in https://github.com/gardener/gardener/pull/5882 by swapping out the mock with an asynchronous fake checker.
